### DISCUSSION TITLE
"Portaling" a writable atom

### DISCUSSION THREAD

**Discussion by tarngerine** [pmndrs/jotai#699] [General]
I've recently enjoyed using React.Portal quite a bit, especially in the context of a design tool with many different types of elements,  with the strategy described in this blog post: https://jjenzz.com/avoid-global-state-colocate

this allows me to structure my code in a way where, e.g. TextElement can render "remotely" into main application layouts based on its state:

```
function TextElement() {
  return <>
    {element.isSelecting && <InspectorPortal><TextInspector /></InspectorPortal>}
  </>
}
```

super handy, quick for prototyping, no need for a big `switch/case` statement inside Inspector that I have to manage

---

so I have run into a situation with jotai where this pattern could make sense? made a proof of concept (linked below) but just wanted to share here to get @dai-shi 's thoughts? https://codesandbox.io/s/headless-smoke-s5viy?file=/src/App.tsx

example problem:
- i have a bunch of elements, e.g. Text, Polygon. Each one has slightly different "deselection" logic
- i have a central writable atom, selectElementAtom, which accepts a string ID or null (for deselection)
- how might we do a similar "portal" pattern from each element, where when it's selected, we "portal"/replace a central writable atom with a writable atom for the specific element type

API might look like

```
// central deselect atom that's "portaled" into
const deselectAtom = atom<WritableAtom<any, any> | null>(null)

// utility hook for portaling a writable atom into deselect
function useAtomPortal(isEnabled, targetAtom, sourceAtom) {
  const setTarget = useUpdateAtom(targetAtom)
  useEffect(() => {
    if (isEnabled) setTarget(sourceAtom)
  }, [isEnabled, sourceAtom])
}

// Text element-only deselect atom, needs to be an atom bc it interacts with other atoms via get/set
const deselectTextAtom = atom(null, (get, set) => { ...do stuff with get, set });

// conditionally portal this atom
function TextElement() {
  useAtomPortal(element.isSelecting, deselectAtom, deselectTextAtom)
}

// main selectElement writable atom we have, which can be called from anywhere
const selectElementAtom = atom(null, (get, set, id: id | null) => {
  // deselect element
  if (id === null) {
    // now element specific logic
    const portaledDeselectAtom = get(deselectAtom)
    if(portaledDeselectAtom) set(portaledDeselectAtom) // calls the portaled function if it exists
  }
  // do other stuff...
});
```

i'm aware i could do this with a switch/case in a central deselect writable atom, but just curious if there's an interesting pattern here. is there a better way i could do something like this?

here's a codesandbox of this in action: https://codesandbox.io/s/headless-smoke-s5viy?file=/src/App.tsx

**Comment by dai-shi** [comment]
At a glance, I wasn't able to get the point, but the codesandbox example helped. (Does "Deselect" button work as expected?)

Not sure if I understand fully, but I guess I like the idea generally:
- Putting an atom config into another atom is a valid pattern. I do a lot for advanced use cases.
- The names are a little confusing, could be improved.
- A little concern with `useEffect`, which is not called immediately. There might be a better way like to wrap the selection callback.

Overall, it looks interesting. Nothing seems wrong from jotai lib perspective, and it can be a powerful practice.

**Comment by tarngerine** [comment]
1. yes, the deselect button is working properly in the codesandbox — expected: when Text element is selected, deselect should run the deselectTextAtom writable atom and log something about "text", similar with Rect selected. with nothing selected nothing logs.
2. do you have other examples where you put atom configs into another one? i'd love to study how you've done that?
3. agreed names are not great, def a weakpoint i have
4. didn't know this about useEffect — could you show me a bit of pseudo code by what "wrap the selection callback" means?

thank you for your feedbacks!

**Comment by dai-shi** [comment]
>     2. do you have other examples where you put atom configs into another one? i'd love to study how you've done that?

https://docs.pmnd.rs/jotai/guides/atoms-in-atom

>     4. didn't know this about useEffect — could you show me a bit of pseudo code by what "wrap the selection callback" means?

```jsx
function TextElement() {
  const [selection, setSelection] = useAtom(selectionAtom);
  const wrap = useAtomPortal(selection === "text", portalAtom, deselectTextAtom); // no useEffect

  return (
    <>
      Text element {selection === "text" ? "selected" : "not selected"}
      <button onClick={() => wrap(setSelection)("text")}>Select text</button>
    </>
  );
}
```

not sure how it works, just something without useEffect, which mean you invoke in the callback somehow.
